### DISCUSSION TITLE
Virtual join

### DISCUSSION THREAD

**Discussion by johnlepikhin** [diesel-rs/diesel#4125] [Q&A] [answered]
Hi there,

I have a code like:

```rust
let results = if true {
  // Select row from tbl1 and one optional field from tbl2
  tbl1
    // ... Skipped many joins, filters, etc
    .inner_join(tbl2.on(tbl2::id.eq(tbl1::id)))
    .select(tbl1::all_columns, tbl2::some_optional_field)
    .load()
} else {
  // Select from tbl1 only, and then always add None as the placeholder for tbl2::some_optional_field
  tbl1
    // ... Skipped many joins, filters, etc. All are the same as in the first condition branch
    .select(tbl1::all_columns)
    .load()
    .map(|row| (row, None))
}
```
Looks like boilerplate could be reduced with boxed expression:

```rust
let mut query = tbl1
    // ... Skipped many joins, filters, etc
    .into_boxed();

let results = if true {
  query = query
    .inner_join(tbl2.on(tbl2::id.eq(tbl1::id)))
    .select(tbl1::all_columns, tbl2::some_optional_field);
} else {
  query = query
    .select(tbl1::all_columns, None); // <- KEY IDEA
}
```
In other words, I want to add constant None into select() and satisfy type checker on missing join in the second branch. Any ideas how to implement this?

EDIT: As a last resort, I can add a join to the second query, so the main question is how to add constant None into select().

**Comment by weiznich** [comment, marked as answer]
> so the main question is how to add constant None into select()

You want to have a look at the [`IntoSql::into_sql`](https://docs.diesel.rs/2.2.x/diesel/expression/trait.IntoSql.html#method.into_sql) method for that.

Otherwise speaking: Diesel (or better the built-in DSL) always want's to know the from clause (including joins) of your query to statically check whether a certain expression is valid in your `select` or any other clause. So you cannot assign different queries with different from clauses to the same variable. What you can easily do is construct the common part outside of your if statement and then only apply the parts that are different + the load in each branch. That should allow skipping all the many joins in your example.
